Observe analytic solutions if available

nilsvu opened the pull request:
## Proposed changes

`ObserveFields` and `ObserveErrorNorms` now support checking at runtime whether or not an analytic solution is available. The check is disabled when no analytic solution tensors are given at compile-time. This is in preparation for #2603, i.e. to allow factory-creating the elliptic solver background (analytic solution or data).

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
- If you retrieved analytic solutions from the DataBox as `::Tags::Analytic<Tag>`, update your code to retrieve `::Tags::AnalyticSolutionsBase` instead. It currently always holds a `Variables` from which you can retrieve the analytic solution fields as `::Tags::Analytic<Tag>` the same way you did before. In the future it may hold a `std::optional<Variables>` in cases where it is not clear at compile-time that analytic solutions are available.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu commented:
@moxcodes this is the same code as in #2603, except I removed the default template parameter that you commented on.

moxcodes commented on src/Elliptic/Actions/InitializeAnalyticSolution.hpp:
```diff
@@ -32,47 +32,52 @@ namespace Actions {
  */
-template <typename AnalyticSolutionTag, typename AnalyticSolutionFields>
+template <typename BackgroundTag, typename AnalyticSolutionFields,
+          typename AnalyticSolutionType>
```
So, is the intended use case for this template parameter that the calling code will identify the 'best' (i.e. most specific) base class that specifies the interface that the calling code requires? Can you add something to the documentation explaining that intended use?

moxcodes commented on src/Elliptic/Actions/InitializeAnalyticSolution.hpp:
```diff
@@ -32,47 +32,52 @@ namespace Actions {
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ElementId<Dim>& /*array_index*/, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    const auto analytic_solution =
```
Sorry, I'm now confused about how this dynamic cast is supposed to work in the generic case. If I'm reading it correctly, you're anticipating that the `BackgroundTag` will hold some factory creatable thing, that may or may not be a derived class of `AnalyticSolutionType`. But then it seems like you have `SomeClass` that inherits both from the base class in `BackgroundTag` and from `AnalyticSolutionType` and you are trying to cast between the two base classes, but I don't think that cast works when there isn't a direct inheritance.

And if there is a direct inheritance, then I think you should be able to just use `std::is_base_of` and avoid the dynamic cast trickery.

Is there a case here that I'm missing?

nilsvu replied:
The inheritance I was aiming for is `SomeClass : AnalyticSolution : Background : AnalyticData` for analytic solutions and `SomeClass : Background : AnalyticData` for analytic data, where the `Background` class may be omitted for systems that don't have extra requirements like a constitutive relation or an EOS.

The `BackgroundTag::type` here is `Background` and the `AnalyticSolutionType` is `AnalyticSolution`. So I don't think `std::is_base_of` would work because I need to determine at runtime which of the two inheritance trees the object is part of.

nilsvu commented:
I'm marking this PR as "draft" for now, because I have to figure out whether or not we'll need this functionality. Thanks for your helpful comments so far @moxcodes. I'd like to get some feedback on my plans for the set of elliptic executables before moving forward, so I'll open an issue and perhaps we can discuss with few people in a call some time.

nilsvu commented:
Removing draft status because we discussed this on 2020-12-21 (see #2695). This is ready for review (again).

nilsvu committed:
Observe analytic solutions if available

Support storing the analytic solutions in a std::optional<Variables>
to signal whether or not solutions are available. The observe events
check availability of analytic solutions.

moxcodes commented on src/Elliptic/Actions/InitializeAnalyticSolution.hpp:
```diff
@@ -26,55 +26,94 @@ struct Inertial;
+ * either case, but it will hold a `std::optional` when
+ * `InitializeOptionalAnalyticSolution` is used. In that case, the analytic
+ * solution is only evaluated and stored in the DataBox if the `BackgroundTag`
+ * holds an `AnalyticSolutionType`.
```
Perhaps note the inheritance hierarchy constraint: "And `AnalyticSolutionType` inherits from `BackgroundTag::type`" or similar.

nilsvu replied:
done

nilsvu commented:
@wthrowe Would you be able to review this, since it touches the observation events?

moxcodes commented:
(any changes to the above please squash immediately)

wthrowe commented on src/Elliptic/Actions/InitializeAnalyticSolution.hpp:
```diff
@@ -26,55 +26,94 @@ struct Inertial;
+ * Use `InitializeAnalyticSolution` if it is clear at compile-time that an
+ * analytic solution is available, and `InitializeOptionalAnalyticSolution` if
+ * that is a runtime decision, e.g. based on a choice in the input file. The
+ * `::Tags::AnalyticSolutionBase` tag can be retrieved from the DataBox in
```
AnalyticSolution**s**Base

wthrowe commented on src/ParallelAlgorithms/Events/ObserveErrorNorms.hpp:
```diff
@@ -125,34 +128,51 @@ class ObserveErrorNorms<ObservationValueTag, tmpl::list<Tensors...>,
+        return;
+      }
+    }
+    const auto& analytic_solutions = [&optional_analytic_solutions]() noexcept {
```
I think you need an explicit `-> decltype(auto)` to avoid the copy.

nilsvu replied:
you're right, good catch!

wthrowe commented on tests/Unit/Elliptic/Actions/Test_InitializeAnalyticSolution.cpp:
```diff
@@ -25,87 +29,144 @@ struct ScalarFieldTag : db::SimpleTag {
+
+struct AnalyticSolutionOrData : PUP::able {
+  using creatable_classes =
+      tmpl::list<AnalyticSolutionOrData, AnalyticSolution>;
```
These classes aren't creatable.  Do the registration manually, instead of pretending to go through a factory.

wthrowe commented on tests/Unit/Elliptic/Actions/Test_InitializeAnalyticSolution.cpp:
```diff
@@ -25,87 +29,144 @@ struct ScalarFieldTag : db::SimpleTag {
+  explicit AnalyticSolutionOrData(CkMigrateMessage* m) : PUP::able(m) {}
+  // For some reason this fails to compile if replaced by just `PUPable_decl`.
+  // Upstream issue: https://github.com/UIUC-PPL/charm/issues/3235
+  WRAPPED_PUPable_decl_template(AnalyticSolutionOrData);
```
You need to pair your `PUPable_decl` with a `PUPable_def` outside the class.

nilsvu replied:
Aha, Thanks!

(These PUPable macros have been a major headache for me.. On multiple occasions now have I spent way too much time trying to figure out the correct combination of macros needed to make the compiler, linker and runtime happy..)

wthrowe commented:
Squash.

nilsvu commented:
@moxcodes could you take another look at this one?

nilsdeppe commented on src/Evolution/ComputeTags.hpp:
```diff
@@ -21,12 +21,9 @@ namespace Tags {
-                                  ::Tags::Variables<AnalyticFieldsTagList>>;
+struct AnalyticCompute : ::Tags::AnalyticSolutions<AnalyticFieldsTagList>,
+                         db::ComputeTag {
+  using base = ::Tags::AnalyticSolutions<AnalyticFieldsTagList>;
```
Just so I understand the plan: this will ultimately return a `std::optional<Variables>`, grabbing `Tags::AnalyticSolutionOrData` (or whatever it's called), checking if the thing is an analytic solution or not, and then set the `std::optional<Variables>` appropriately. Is that correct?

nilsvu replied:
Yep!
